### ISSUE TITLE
🐛 BUG: port=0 regression in wrangler@3.18.0

### ISSUE THREAD

**Issue by dario-piotrowicz** [cloudflare/workers-sdk#4563]
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

wrangler 3.18.0

### What version of Node are you using?

18

### What operating system are you using?

Mac

### Describe the Bug

In `wrangler@3.17.0` and prior it was possible to run `wrangler dev --port=0`, listen to its `DEV_SERVER_READY` ipc message and fetch from the provided address, it seems that however this has broken in `wrangler@3.18.0` as now `fetch`es to the address result in a `ECONNREFUSED` error

Some notes on this:
 - The same applies to `wrangler pages dev` as well
 - If no `--port=0` is specified everything works as expected
 - if a valid non-zero port is specified everything works as expected
 - only `fetch`es seem to be effected, I can still navigate to the address with the browser and get the correct result

### Please provide a link to a minimal reproduction

https://github.com/dario-piotrowicz/wrangler-dev-local-port-0-issue-reproduction/

**Comment by petebacondarwin** [comment]
Some digging into this.  It appears that when using `wrangler dev --port=0`, we execute workerd with something like this:

```sh
workerd serve --binary --experimental --socket-addr=entry=localhost:0 ...
```

Note that we are sending localhost:0 as the socket it should listen on.

In workerd `localhost` is being resolved to two IP addresses: `127.0.0.1` and `::1`. And since we passed in port 0, it is asking the OS to provide a port for these addresses. Unfortunately, it appears to do this separately for each address so it gets two ports and listens to one port for each address. You can see this from an `lsof`:

```
workerd   72782 pbacondarwin   14u  IPv4 0x1e4003578d31d345      0t0  TCP 127.0.0.1:56632 (LISTEN)
workerd   72782 pbacondarwin   15u  IPv6 0x1e400365e67f3145      0t0  TCP [::1]:56633 (LISTEN)
```

Note that localhost IPv4 is listening on 56632 and IPv6 is listening on 56633.

Wrangler only knows about the first port, so it tells the user that it can make fetch requests on `localhost:56632`. Unfortunately node fetch will convert `localhost` to `::1` which means that there is a mismatch between the IP and the port (e.g. `[::1]:56632`) which workerd is not listening on.

**Comment by dario-piotrowicz** [comment]
Closing this issue as the bug seems to have been fixed as of `3.21.0` (I'm not really sure what fixed it though 😕)
